text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/MiniMessageParser.java

```diff
@@ -189,39 +196,46 @@ private void processTokens(final @NotNull StringBuilder sb, final @NotNull Strin
   }
 
   @NotNull Component treeToComponent(final @NotNull ElementNode node, final @NotNull ContextImpl context) {
-    Component comp;
-    Transformation transformation = null;
+    Component comp = Component.empty();
+    Tag tag = null;
     if (node instanceof ValueNode) {
       comp = Component.text(((ValueNode) node).value());
     } else if (node instanceof TagNode) {
-      final TagNode tag = (TagNode) node;
+      final TagNode tagNode = (TagNode) node;
 
-      transformation = tag.transformation();
+      tag = tagNode.tag();
 
       // special case for gradient and stuff
-      if (transformation instanceof Modifying) {
-        final Modifying modTransformation = (Modifying) transformation;
+      if (tag instanceof Modifying) {
+        final Modifying modTransformation = (Modifying) tag;
 
         // first walk the tree
-        final LinkedList<ElementNode> toVisit = new LinkedList<>(node.children());
+        final LinkedList<ElementNode> toVisit = new LinkedList<>(node.unsafeChildren());
         while (!toVisit.isEmpty()) {
           final ElementNode curr = toVisit.removeFirst();
           modTransformation.visit(curr);
-          toVisit.addAll(0, curr.children());
+          toVisit.addAll(0, curr.unsafeChildren());
```

Comment:
Why do these need to use **unsafe** children? I tested with regular children and couldn't see any difference?

Reply:
no point in constructing an unmodifiable wrapper collection for every call while processing the transformation, when this is internal use we can control